feat: Allow remote editor for WPCOM Simple sites

dcalhoun opened the pull request:
Allow WPCOM Simple sites to use the experimental remote, site-specific editor. Close CMM-222.

Related: 

- CMM-222-linear-issue
- 177481-ghe-Automattic/wpcom
- https://github.com/Automattic/jetpack/pull/42314
- https://github.com/wordpress-mobile/GutenbergKit/pull/109
- https://github.com/wordpress-mobile/WordPress-Android/pull/21773

To test:

1. Navigate to Me → App Settings → Experimental Features and enable `Experimental Block Editor Plugins`
1. Open a post on a WPCOM Simple site. 
1. Insert non-core Jetpacks blocks. 
1. Verify they function as expected. 

## Regression Notes
1. Potential unintended areas of impact
    The local, bundled experimental editor. 
6. What I did to test those areas of impact (or what existing automated tests I relied on)
    Manually smoke tested. 
7. What automated tests I added (or what prevented me from doing so)
    Deemed unnecessary for this experimental feature. 

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

dcalhoun committed:
feat: Enable plugins for WPCOM sites

Adjust strategy to shorten time-to-ship necessary REST API endpoints.

dcalhoun committed:
feat: Provide all REST API namespace patterns to editor

Ensures the editor can detect namespace paths regardless of which
pattern is used.

dcalhoun committed:
feat: Provide editor namespace excluded paths

Some REST APIs are site-specific and require a site namespace, others
are global.

dcalhoun committed:
feat: Set site type global

Enables Jetpack plugin blocks to accurately detect Jetpack's
"connection" status.

dangermattic commented:
<!--
  0 Errors
  2 Warnings: <code>Modules/Package.swift</c...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages as appropriate to your project setup (e.g. in Xcode or by running <code>swift package resolve</code>).</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dcalhoun committed:
feat: Expose experimental editor plugins feature

dcalhoun committed:
feat: Limit experimental editor plugin support to WPCOM Simple sites

Until we can authenticate via application passwords, Jetpack-connected
sites fail to authenticate API requests originating from plugin blocks.
E.g., Jetpack blocks currently make failing requests to the site domain
rather than the WPCOM API.

dcalhoun committed:
build: Update GutenbergKit ref

dcalhoun committed:
build: Update GutenbergKit ref

dcalhoun committed:
feat: Control GutenbergKit plugin support availability

Allow remotely disabling this experimental feature while the API
endpoint is deployed or during a breakage.

dcalhoun commented on WordPress/Classes/Utility/BuildInformation/RemoteFeatureFlag.swift:
```diff
@@ -205,6 +210,8 @@ enum RemoteFeatureFlag: Int, CaseIterable {
         case .dotComWebLogin:
             return "Log in to WordPress.com from web browser"
+        case .newGutenbergPluginsAvailability:
+            return "Experimental Block Editor Plugins Availability"
```
Used to ensure we can remotely control whether the plugin feature is available—for API endpoint rollout or breakages.

dcalhoun committed:
build: Update GutenbergKit ref

dcalhoun committed:
refactor: Update remote feature flag key

Match the latest server implementation.

kean commented:
I think I followed the steps closely enough, but I'm getting the following issue. I haven't debugged it, but I don't see errors in the in-app logs. Is there something else I need to do? How can I debug it?

<img src="https://github.com/user-attachments/assets/8e6e9b98-f1fb-44b1-920a-6daf122040f8" width=320px>

dcalhoun commented:
> I think I followed the steps closely enough, but I'm getting the following issue. I haven't debugged it, but I don't see errors in the in-app logs. Is there something else I need to do? 

* Have you applied the testing instructions setup from https://github.com/Automattic/jetpack/pull/42314 to your sandbox? 
* Are you using a simulator or physical device? The latter requires proxying your device traffic through your development machine so that request are made to your sandbox. 

> How can I debug it?

I would monitor your network traffic to observe the response from the editor assets endpoint. 

If that is successful, then debugging further would likely require building locally and modifying the JavaScript to log additional details rather than redirecting to the local, bundled editor—i.e., the yellow notice displayed. If it comes to this, I'm happy to pair on debugging the error.

kean reviewed (approved):
I'll just have to approve it based on the code review. The native part looks good, except that I would recommend to use local flags for experimental features.

We have a pretty extreme timeline for another project, and I don't have a sandbox configured that I can use for testing.

dcalhoun commented:
> The native part looks good, except that I would recommend to use local flags for experimental features.

@kean the remote feature flag added both (A) toggles the visibility/availability of the experimental feature toggle and (B) gates the feature itself (i.e., can override a user enabling the experimental feature itself). My rationale was that a remote feature flag allows us to remotely disable the feature if it proved unstable or problematic. Do you feel this is unnecessary?

kean commented:
> remotely disable the feature if it proved unstable or problematic

I'm not sure it's necessary considering that it's an experimental feature. It doesn't hurt adding a remote flag. I just left it as a suggestion.

dcalhoun committed:
feat: Remove remote feature flag

This controlled the availability of the experimental feature toggle and
also could override the feature flag status. This additional remote
feature flag was deemed unnecessary for an experimental feature.

https://github.com/wordpress-mobile/WordPress-iOS/pull/24170#issuecomment-2769617420

dcalhoun commented:
FYI: the related server changes adding the REST API endpoint were deployed, so there is no longer a need to route device traffic to a sandbox server. I updated the PR description accordingly.

dcalhoun committed:
build: Update GutenbergKit ref
